src/charm.py

```diff
@@ -357,21 +358,51 @@ def _are_db_relations_ready(self) -> bool:
             return False
         return True
 
-    def _config_changed(self, event: HookEvent) -> None:
-        """Configure pod using pebble and layer generated from config.
+    def _set_up_discourse(self, event: HookEvent) -> None:
+        """Run Discourse migrations and recompile assets.
 
         Args:
             event: Event triggering the handler.
         """
-        self.model.unit.status = MaintenanceStatus("Configuring service")
-        if not self._are_db_relations_ready():
+        container = self.unit.get_container(SERVICE_NAME)
+        if not self._are_db_relations_ready() or not container.can_connect():
             event.defer()
             return
+        env_settings = self._create_discourse_environment_settings()
+        try:
+            if self.model.unit.is_leader():
+                self.model.unit.status = MaintenanceStatus("Executing migrations")
+                process = container.exec(
+                    [f"{DISCOURSE_PATH}/bin/bundle", "exec", "rake", "--trace", "db:migrate"],
+                    environment=env_settings,
+                    working_dir=DISCOURSE_PATH,
+                    user="discourse",
+                )
+                process.wait_output()
+            self.model.unit.status = MaintenanceStatus("Compiling assets")
+            process = container.exec(
+                [f"{DISCOURSE_PATH}/bin/bundle", "exec", "rake", "assets:precompile"],
+                environment=env_settings,
+                working_dir=DISCOURSE_PATH,
+                user="discourse",
+            )
```

Comment:
Is the install/upgrade even triggered if the container is rescheduled ?
A reschedule will "lose" the precompiled asset, if the event is not re-triggered we'll get an issue with this change

Reply:
Also.. shouldn't all units run the precompile assets ? I don't think they're in a folder shared by all units, therefore each unit should have it local precompiled assets no ?

Reply:
Exactly. That's the current behaviour, but the migrations are only run in the leader unit

Reply:
Oh yes indeed.
The comment is still valid concerning the "install/upgrade" event even if it seems that the upgrade event is triggered when the container is rescheduled